src/backend/cdb/motion/ic_proxy.h

```diff
@@ -22,8 +22,13 @@
 #define IC_PROXY_BACKLOG 1024
 #define IC_PROXY_INVALID_CONTENT ((uint16) -2)
 #define IC_PROXY_INVALID_DBID ((int16) 0)
-#define IC_PROXY_TRESHOLD_PAUSE 4
-#define IC_PROXY_TRESHOLD_RESUME 2
+/* pause the sender when the unack packet increase to this threshold */
+#define IC_PROXY_TRESHOLD_UNACK_PACKET_PAUSE 100
+/* resume the sender when the unack packet reduce to this threshold */
+#define IC_PROXY_TRESHOLD_UNACK_PACKET_RESUME 30
+/* send a ack message after a batch of packets*/
+#define IC_PROXY_PACKET_INTERVAL_OF_ACK_MESSAGE 10
```

Comment:
I have some suggestions on the names, but feel free to keep the current ones, they are good enough.

```suggestion
/* pause the sender when the unack packet increase to this threshold */
#define IC_PROXY_TRESHOLD_PAUSE 100
/* resume the sender when the unack packet reduce to this threshold */
#define IC_PROXY_TRESHOLD_RESUME 30
/* send a ack message after a batch of packets*/
#define IC_PROXY_ACK_INTERVAL 10
```

BTW, the pause threshold, 100, will it be too large in practice?  And do you have some benchmark data on the performance impact?

Reply:
100 is necessary when considering the query latency. I firstly set it to 20 and for the query like `select count(*) from a1,a2 where a1.j=a2.j;`(detail in the beginning of the PR message) which includes a redistributed motion node. the Motion time in explain analyze output is 3 seconds. When change IC_PROXY_TRESHOLD_UNACK_PACKET_PAUSE to 100, it's about 1 second.

The reason is that the ack mechanism is different from the udp. The receiver cannot send ack message immediately, but needs the backend consumed the packets. This leads to the delay of ack.
I think IC_PROXY_TRESHOLD_UNACK_PACKET_PAUSE is a tradeoff between query latency and memory usage. In future, we may adjust this value as a GUC.

Reply:
Makes sense.